Lib/ufo2ft/__init__.py

```diff
@@ -872,6 +928,62 @@ def _compileNeededSources(
 
         # Stick TTFs back into original big DS
         for ttfSource in ttfDesignSpace.sources:
+            if can_optimize_features:
+                originalSources[ttfSource.name] = sourcesByName[ttfSource.name].font
             sourcesByName[ttfSource.name].font = ttfSource.font
 
-    return vfNameToBaseUfo
+    return vfNameToBaseUfo, can_optimize_features, originalSources
+
+
+def compile_all_variable_features(
+    designSpaceDoc, vfNameToTTFont, originalSources, debugFeatureFile=False
+):
+    interpolableSubDocs = [
+        subDoc for _location, subDoc in splitInterpolable(designSpaceDoc)
+    ]
+    for subDoc in interpolableSubDocs:
+        for vfName, vfDoc in splitVariableFonts(subDoc):
+            if vfName not in vfNameToTTFont:
+                continue
+            ttFont = vfNameToTTFont[vfName]
+            # vfDoc is now full of TTFs, create a UFO-sourced equivalent
+            ufoDoc = vfDoc.deepcopyExceptFonts()
+            for ttfSource, ufoSource in zip(vfDoc.sources, ufoDoc.sources):
+                ufoSource.font = originalSources[ttfSource.name]
+            compile_variable_features(ufoDoc, ttFont, debugFeatureFile)
+
+
+def compile_variable_features(designSpaceDoc, ttFont, debugFeatureFile):
+    default_ufo = designSpaceDoc.findDefault().font
+
+    # Delete anything from the UFO glyphset which didn't make it into the font.
+    fontglyphs = ttFont.getGlyphOrder()
+    glyphSet = {g.name: g for g in default_ufo if g.name in fontglyphs}
+
+    # Add anything we added to the TTF without telling the UFO
+    if ".notdef" not in glyphSet:
+        glyphSet[".notdef"] = StubGlyph(".notdef", 0, 0, 0, 0)
+
+    featureCompiler = VariableFeatureCompiler(
+        default_ufo, designSpaceDoc, ttFont=ttFont, glyphSet=glyphSet
+    )
+    featureCompiler.compile()
+
+    if debugFeatureFile:
+        if hasattr(featureCompiler, "writeFeatures"):
+            featureCompiler.writeFeatures(debugFeatureFile)
+
+    # Add back feature variations, as the code above would overwrite them.
+    designSpaceData = varLib.load_designspace(designSpaceDoc)
+    featureTag = designSpaceData.lib.get(
+        varLib.FEAVAR_FEATURETAG_LIB_KEY,
+        "rclt" if designSpaceData.rulesProcessingLast else "rvrn",
+    )
+    if designSpaceData.rules:
+        varLib._add_GSUB_feature_variations(
+            ttFont,
+            designSpaceData.axes,
+            designSpaceData.internal_axis_supports,
+            designSpaceData.rules,
+            featureTag,
+        )
```

Comment:
maybe we should add a new public method to varLib that does just that, instead of using a private method like this. Something that takes a DS and a VF, and translates all the DS rules into equivalent featureVars

Reply:
we should now be able to use the varLib.addGSUBFeatureVariations function that I added with fonttools 4.46: https://github.com/fonttools/fonttools/pull/3362